### ISSUE TITLE
Support for multiple services living in the same process.

### ISSUE THREAD

**Issue by Qrious** [elastic/apm-agent-dotnet#1001]
**Is your feature request related to a problem? Please describe.**
Currently it is no longer possible (to my knowledge, i've looked in the documentation and issues) to override the service name when starting a transaction. This is an issue in our use case. We use an actor based distributed system framework. The key issue here is that we don't have a service per process model, but that multiple services might run in the same process (controlled by hypervisor). As a result, setting an Environment variable is not a viable solution, since that would result in race conditions between multiple services setting/reading the Environment variable.

We believe this is an issue in general for actor based distributed systems (such as Microsoft Service Fabric and Akka.Net).

I do apologize if it still is actually possible to override the service name, but i've not found any way/documentation on how to do this.

**Describe the solution you'd like**
Add an option when creating a transaction to specify the service name.

As a side note, this was possible in earlier versions (I've had seen this working at version 1.0.5).

**Describe alternatives you've considered**
- Using the environment variable (leads to race condition)
- Overriding it using `Custom` context/labels. This did (obviously) not work.

**Additional context**
-

**Comment by gregkalapos** [comment]
Thanks for reporting this @Qrious.

How did you solve this problem in earlier versions? 

I'd like to say that in general we have a single agent per process approach and right now the service name is tied to this single agent version. But I totally see your use case and it makes perfect sense.

One thing which is possible in the APM Server is that an agent can overwrite the service name for a given transaction. So having the single agent per process approach we could still give you an option to override the service name on a per transaction basis.

There is no API for it, but it could be something like this:

```
if(tracer.CurrentTransaction != null)
        // SetService does not exists, but it''d be something like this:
	tracer.CurrentTransaction.SetService(name: "MyServiceName", version: "2.1-beta");
```

In this case you'd need to do this for each transaction.

If there would be such an API, it'd be also possible to use it through the Filter API:

```
Agent.AddFilter(transaction =>
{
	if(IsThisACustomService())
		transaction.SetService(name: "MyServiceName", version: "2.1-beta");
	
	return transaction;
});
```
A filter runs right before the transaction is sent to the APM Server - so if you can isolate your services (that's what the `IsThisACustomService()` simulates) - then this would be a nicer way to override the service name.

If you don't overwrite the service name on a transaction then it'd be the default service name filled by our current default logic which uses the entry assembly name.


These are just my early thoughts - so nothing is final on this. How does this sound? Would that help with your use case?

**Comment by Qrious** [comment]
In a much earlier version (1.0.5) it was possible to do the following:
```csharp
await Agent.Tracer.CaptureTransaction(eventToProcess.EventName, "MyEventType", async (trans) =>
{
        trans.Context.Service.Name = "MyServiceName";
        // More code here
});
```

Since you are basically restoring that API (but making it a bit better by adding a method call), that sounds perfect for me. 

I think in our case I can't easily resolve the service from a filter other than passing it via labels/customs, and reading those values in the filter, but I fully agree that is would be a nice place to modify the servicename for people that are able to easily resolve which service created the transaction.

**Comment by gregkalapos** [comment]
Implemented in #1002.
